packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+import * as asyncHooks from 'async_hooks';
+import { EventEmitter } from 'events';
+import * as shimmer from 'shimmer';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> = [
+  'addListener',
+  'on',
+  'once',
+  'prependListener',
+  'prependOnceListener',
+];
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = {};
+  private current: unknown = null;
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      before: this.before.bind(this),
+      destroy: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.current;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: T
+  ): ReturnType<T> {
+    const oldCurrent = this.current;
+    this.current = scope;
+    const res = fn();
+    this.current = oldCurrent;
+    return res;
+  }
+
+  bind<T>(target: T, scope?: unknown): T {
+    // if no specific scope to propagate is give, we use the current one
+    if (!scope) {
+      scope = this.current;
+    }
+    if (target instanceof EventEmitter) {
+      return this.bindEventEmitter(target, scope);
+    } else if (typeof target === 'function') {
+      return this.bindFunction(target, scope);
+    }
+    return target;
+  }
+
+  enable(): this {
+    this.asyncHook.enable();
+    return this;
+  }
+
+  disable(): this {
+    this.asyncHook.disable();
+    this.scopes = {};
+    this.current = null;
+    return this;
+  }
+
+  private bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const boundContext = scope;
+    const contextWrapper = function(this: {}) {
+      const oldContext = manager.current;
+      manager.current = boundContext;
+      const res = target.apply(this, arguments);
+      manager.current = oldContext;
+      return res;
+    };
+    Object.defineProperty(contextWrapper, 'length', {
```

Comment:
Defined properties have a high cost. Could cases under a certain amount be handled directly instead? For example under 10 properties it could be handled with functions like `function(a, b, c, etc) {}`.

Reply:
I don't see how you suggest to operate here, could you give me an example ?

Reply:
`eval` could be used but that would probably trigger a lot of red flags.

Another option is to do something like this:

```js
switch (fn.length) {
  case 1:
    return function (a) {
      fn.apply(this, arguments)
    }
  case 2:
    return function (a, b) {
      fn.apply(this, arguments)
    }
  // ... and so on
  default:
    // here either return without changing the length, or change the length with `defineProperty`
}
```

Reply:
Do you have any benchmark around that shows a significant impact ? I would prefer keeping the code simple

Reply:
https://github.com/open-telemetry/opentelemetry-js/issues/127